### PR TITLE
Devdocs: WordPress components gallery

### PR DESCRIPTION
**Pull request by saramarcondes**
#### Changes proposed in this Pull Request

* Add a new WordPress components gallery page to the devdocs

Captura de Tela 2020-09-11 às 10 32 25

#### Testing instructions

* Using an environment that has devdocs enabled, navigate to the new WordPress components gallery and check that each component is represented.

Note that some components are represented through the building of the page itself, for example, `Card` et co as well as the `Flex` module were used to build the page, so I didn't feel it was necessary to include them as distinct examples. Actually using them is probably the best way to show that they work.

### CODE REVIEW

File: client/devdocs/design/wordpress-components-gallery/autocomplete.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+					<div
+						className="wordpress-components-gallery__autocomplete-editable"
+						contentEditable
+						onKeyDown={ onKeyDown }
+						suppressContentEditableWarning
+						aria-autocomplete="list"
+						aria-expanded={ isExpanded }
+						aria-owns={ listBoxId }
+						aria-activedescendant={ activeId }
+					></div>
```
Comment (tyxla) [inline client/devdocs/design/wordpress-components-gallery/autocomplete.tsx:58]:
```suggestion
					/>
```

File: client/devdocs/design/wordpress-components-gallery/autocomplete.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+					/* eslint-enable */
+				) }
+			</Autocomplete>
+			<p>Type ~ for triggering the autocomplete.</p>
```
Comment (tyxla) [inline client/devdocs/design/wordpress-components-gallery/autocomplete.tsx:62]: Not sure why I'm unable to make this work by typing `~` 😉

File: client/devdocs/design/wordpress-components-gallery/angle-picker-control.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ * WordPress dependencies
+ */
+import { AnglePickerControl } from '@wordpress/components';
+
+const Example = () => {
+	const [ angle, setAngle ] = useState();
+	return <AnglePickerControl value={ angle } onChange={ setAngle } />;
+};
+
+export default Example;
```
Comment (tyxla) [inline client/devdocs/design/wordpress-components-gallery/angle-picker-control.tsx:16]: I've noticed all component examples are named `Example`. Should we have a different name for each component, like component examples in Calypso do? So we'd have `AnglePickerControlExample` here, for example.

File: client/devdocs/design/wordpress-components-gallery/button.scss [hunk 4]
```diff
@@ -0,0 +1,10 @@
+.wordpress-components-gallery-buttons {
+    &__container {
+        display: flex;
+        
+        & > * {
```
Comment (tyxla) [inline client/devdocs/design/wordpress-components-gallery/button.scss:5]: I wonder if we meant to make this a tad more specific.

File: package.json [hunk 5]
```diff
@@ -211,7 +211,7 @@
 		"@types/wordpress__api-fetch": "^3.2.2",
 		"@types/wordpress__block-editor": "^2.2.8",
 		"@types/wordpress__block-library": "^2.6.0",
-		"@types/wordpress__components": "^8.5.4",
+		"@types/wordpress__components": "^9.8.0",
```
Comment (tyxla) [inline package.json:214]: Any concern from the gutenboarding point of view from updating this? Should we test anything before shipping? cc @Automattic/luna

**Review by tyxla — COMMENTED** [review]
Great work @saramarcondes!

I think we should ship this as soon as possible, and it will help inform our decisions for any future work, related to global style migrations. Any styling problems or visual glitches should be subject to future work. We can also iterate on the particular examples in future PRs as well.

I've left some nits here and there, but nothing I feel is particularly blocking. 

One thing I wanted to ask you was the bundle size changes - do you have an idea why there are changes to sections that don't seem related? Do we just need to recalculate them?

Again, great to see this in Calypso! 🎉

File: package.json (hunk 5 above)
Comment (yansern) [reply package.json:214]: I think a quick smoke test is to see if anything breaks on `/new`. Current link: https://hash-829c18b8bcfddd5cf96dd3c4bfb03dd0b6df309f.calypso.live/new

Comment (simison) [reply package.json:214]: Type issues should fail loudly in CI anyway, no?

File: client/devdocs/design/wordpress-components-gallery/button.scss [hunk 6]
```diff
@@ -0,0 +1,10 @@
+.wordpress-components-gallery-buttons {
```
[the remark this answers is not in the document]
Comment (saramarcondes) [reply client/devdocs/design/wordpress-components-gallery/button.scss:1]: Ah no, it's meant to be deleted as I just ended up using the `Flex` and `FlexItem` components instead. Thanks for catching that!

File: client/devdocs/design/wordpress-components-gallery/autocomplete.tsx (hunk 2 above)
Comment (saramarcondes) [reply client/devdocs/design/wordpress-components-gallery/autocomplete.tsx:62]: Yes, it doesn't seem to work, I think it's potentially one of the things that is broken 😱

File: client/devdocs/design/wordpress-components-gallery/angle-picker-control.tsx (hunk 3 above)
Comment (saramarcondes) [reply client/devdocs/design/wordpress-components-gallery/angle-picker-control.tsx:16]: I'll rename the default exports for each one to match the file name.

File: package.json (hunk 5 above)
Comment (saramarcondes) [reply package.json:214]: It shouldn't be necessary to smoke test anything given it's just type definition updates.

File: client/devdocs/design/wordpress-components-gallery/autocomplete.tsx (hunk 2 above)
Comment (tyxla) [reply client/devdocs/design/wordpress-components-gallery/autocomplete.tsx:62]: Maybe it's something related to how its popover works? But we can investigate in another PR, I don't consider that a blocker 😉

**Comment by saramarcondes** [comment]
> One thing I wanted to ask you was the bundle size changes - do you have an idea why there are changes to sections that don't seem related? Do we just need to recalculate them?

I have no idea why the bundle size changed for Gutenboarding in this PR. The only thing I can think is that maybe because we're importing all the WP components that some tree-shaking isn't working as expected anymore? Maybe @sgomes would have some insight for how to investigate.

**Comment by saramarcondes** [comment]
Putting everything behind async load marginally improved some things but increased the manifest size by a similar amount that it reduced the Gutenboarding size... and it didn't even fix the problem with Gutenboarding getting bigger.

I don't think this is the right solution.

**Comment by tyxla** [comment]
> Putting everything behind async load marginally improved some things but increased the manifest size by a similar amount that it reduced the Gutenboarding size... and it didn't even fix the problem with Gutenboarding getting bigger.
> 
> I don't think this is the right solution.

Yeah, I'd doubt that would be a good solution. A component like this would generally end up being in the `entry-main` chunk because it would be used in multiple sections and locations throughout the app. Or maybe we could work on an exception for `@wordpress/components` so they would end up in their own chunk?

**Comment by jsnajdr** [comment]
I did some research about the `entry-gutenboarding` chunk group that grew 3.4% parsed size and 1.7% gzipped size after introducing the WordPress components gallery. The difference between the async-load-every-component version and the "normal" one is negligible.

The entire `entry-gutenboarding` chunk group uses 4 shared chunks, shared with other chunk groups. The number of the chunks is the same before and after, but webpack reshuffled the modules a bit. That's expected, as a lot of code from the `@wordpress/*` namespace is used in a brand new place.

The `entry-gutenboarding` chunk group contains exactly the same number of modules both before and after, namely 1865. Some modules were moved from shared chunks to the non-shared `entry-gutenboarding` chunk or vice versa, but none were added or removed.

I think the entire growth can be attributed to the module concatenation optimization being less efficient after the WordPress comonent gallery is added. For a module to be concatenate-able, it must be used only once, and not referenced from any other chunk. It's just like inlining functions -- there can't be a pointer to an inline function, and using a pointer to a function prevents it to be optimized by inlining.

With the WordPress component gallery, many components that were until now used in only one chunk are now used in two. Therefore preventing their concatenation.

I see two solutions:
- just accept the growth and live with it, treating it as a quirk of an optimization algorithm that we can't influence
- build the devdocs section as a separate webpack compilation, not sharing anything with the main app. By nature, devdocs use a lot of components that are not very related to each other, and would otherwise be used only once in the app. They probably cause more de-optimizations like this.

**Comment by simison** [comment]
>  With the WordPress component gallery, many components that were until now used in only one chunk are now used in two. Therefore preventing their concatenation.

Thanks for clear explanation @jsnajdr !

>  just accept the growth and live with it, treating it as a quirk of an optimization algorithm that we can't influence

I'd be fine with this (cc @Automattic/luna). Considering that we want to move towards using `@wordpress/*` packages across the board, we'll have more shared usage ahead and cannot really avoid this anyway.

**Comment by saramarcondes** [comment]
> The difference between the async-load-every-component version and the "normal" one is negligible.

Given that I think the "normal" one is preferrable.

Thanks for the explanation, Jarda. It is super helpful.

**Comment by griffbrad** [comment]
Thanks for the very detailed explanation about the webpack optimization issue @jsnajdr. 

> * just accept the growth and live with it, treating it as a quirk of an optimization algorithm that we can't influence

I agree that accepting the cost here is fine. The hope is that we’re going to use these components throughout Calypso so they’ll end up being shared by every entrypoint anyway. 

> * build the devdocs section as a separate webpack compilation, not sharing anything with the main app. By nature, devdocs use a lot of components that are not very related to each other, and would otherwise be used only once in the app. They probably cause more de-optimizations like this.

This is still probably worth pursuing. devdocs includes many large dependencies that will never be used elsewhere in Calypso. For example, react-live is over 2 MB in size and only useful in the context of devdocs.

**Comment by tyxla** [comment]
I've rebased this to take advantage of the `@wordpress/components` stylesheet import that we did in #45724.

**Comment by tyxla** [comment]
FWIW, lint error and TC build failure are expected - it's happening because we're still using the `url` package in devdocs. I'm fixing that in #45898.

**Comment by tyxla** [comment]
Just like @griffbrad, I agree both of the solutions @jsnajdr suggested make sense - both accepting the trade-offs and considering moving devdocs to either its own entry point or a totally separate app.

I'm actually happy to ship this already since it will be very helpful as we're working on the global styles migration and we get to use more and more `@wordpress/components` in Calypso.

Does anyone object to shipping it as-is? cc @Automattic/luna @Automattic/team-calypso

**Comment by simison** [comment]
@tyxla I'll leave it to your team to decide, thanks for the ping. 👍 

Really looking forward to see where this brings us!

**Review by griffbrad — APPROVED** [review]
I agree. Let’s ship it.

**Comment by tyxla** [comment]
Thanks, @saramarcondes, for the brilliant work here! Also thanks everyone for the feedback! Shipping.

**Pull request merged**